File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
    answer = []
    
    minq = []
    maxq = []
    
    for op in operations:
        cmd, num = op.split()
        num = int(num)
        
        if cmd == 'I':
            heapq.heappush(minq, num)
            heapq.heappush(maxq, -num)
        elif minq and maxq:
            if num == -1:
                heapq.heappop(minq)
                del maxq[-1]
            else:
                heapq.heappop(maxq)
                del minq[-1]
    
    return [-maxq[0], minq[0]] if minq and maxq else [0,0]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <set>

using namespace std;

vector<int> solution(vector<string> operations) {
    vector<int> answer;
    multiset<int> ms; // 오름차순으로 자동 정렬되는 집합
    
    for (auto op :operations) {
        char cmd = op[0];
        int num = stoi(op.substr(2));
        
        if (cmd == 'I') {
            ms.insert(num);
        } else {
            if (ms.empty()) continue;
            
            if (num == -1) {
                ms.erase(ms.begin());
            } else {
                auto it = prev(ms.end());
                ms.erase(it);
            }
        }
    }
    
    if (ms.empty()) return {0, 0};
    
    return {*prev(ms.end()), *ms.begin()};
}